libraries/reliable_amqp_publisher/reliable_amqp_publisher.cpp

```diff
@@ -0,0 +1,247 @@
+   reliable_amqp_publisher_impl(const std::string url, const std::string exchange, const std::string routing_key, const boost::filesystem::path unconfirmed_path, const std::optional<std::string> message_id) :
+     data_file_path(unconfirmed_path), exchange(exchange), routing_key(routing_key), message_id(message_id) {
+      std::string host = "localhost", user = "guest", pass = "guest", path;
+      uint16_t port = 5672;
+      fc::url parsed_url(url);
+      FC_ASSERT(parsed_url.proto() == "amqp", "Only amqp:// URIs are supported for AMQP addresses");
+      if(parsed_url.host() && parsed_url.host()->size())
+         host = *parsed_url.host();
+      if(parsed_url.user())
+         user = *parsed_url.user();
+      if(parsed_url.pass())
+         pass = *parsed_url.pass();
+      if(parsed_url.port())
+         port = *parsed_url.port();
+      if(parsed_url.path())
+         path = parsed_url.path()->string();
+
+      std::stringstream hostss;
+      hostss << "amqp://" << user << ":" << pass << "@" << host << ":" << port << "/" << path;
+      amqp_address = std::make_unique<AMQP::Address>(hostss.str());
+
+      if(boost::filesystem::exists(data_file_path)) {
+         try {
+            fc::datastream<fc::cfile> file;
+            file.set_file_path(data_file_path);
+            file.open("rb");
+            fc::raw::unpack(file, message_deque);
+         } FC_RETHROW_EXCEPTIONS(error, "Failed to load previously unconfirmed AMQP messages from ${f}", ("f", (fc::path)data_file_path));
+      }
+      else {
+         boost::filesystem::ofstream o(data_file_path);
+         FC_ASSERT(o.good(), "Failed to create unconfirmed AMQP message file at ${f}", ("f", (fc::path)data_file_path));
+      }
+      boost::system::error_code ec;
+      boost::filesystem::remove(data_file_path, ec);
+
+      thread = std::thread([this]() {
+         try {
+            fc::set_os_thread_name("amqp");
+            retry_connection();
+            ctx.run();
+         }
+         catch(...) {
+            elog("Exception escaped from AMQP connection ${s} publishing to \"${e}\", this connection is permanently disabled now",
+                 ("s", (std::string)*amqp_address)("e", this->exchange));
+         }
+      });
+   }
+
+   ~reliable_amqp_publisher_impl() {
+      stopping = true;
+
+      //drain any remaining items on queue
+      std::promise<void> shutdown_promise;
+      ctx.post([&]() {
+         shutdown_promise.set_value();
+      });
+      shutdown_promise.get_future().wait();
+
+      ctx.stop();
+      thread.join();
+
+      //if a message is in flight, let it try to run to completion
+      if(in_flight) {
+         auto until = std::chrono::steady_clock::now() + std::chrono::seconds(5);
+         ctx.restart();
+         while(in_flight && ctx.run_one_until(until));
+      }
+
+      if(message_deque.size()) {
+         try {
+            boost::filesystem::create_directories(data_file_path.parent_path());
```

Comment:
If the directory doesn't exist, won't the create check in the constructor fail (assuming it wasn't deleted while we were running)?

Reply:
Probably left over cruft from when I changed the location of creating these directories from previous feedback https://github.com/EOSIO/eos/pull/9121#discussion_r428764014

No real problem to only mkdir once at start up I guess